# project_2. Анализ вакансий на hh.ru

In [41]:
! pip install psycopg2


[notice] A new release of pip is available: 24.0 -> 24.1.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [42]:
import psycopg2
import pandas as pd

In [44]:

connection = psycopg2.connect(
   dbname=DBNAME,
   user=USER,
   host=HOST,
   password=PASSWORD,
   port=PORT
)

In [45]:
query = f'''select *
           from vacancies
           
        '''
# лучше использовать f-строки — так удобнее передавать параметры

In [46]:
df = pd.read_sql_query(query, connection)
df

C:\Users\Polta\AppData\Local\Temp\ipykernel_15108\137330334.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


id                                               name  \
0      55312386                                Компьютерный Мастер   
1      55843014                            Системный администратор   
2      54525964                      Lead Java Developer to Poland   
3      54525965                      Lead Java Developer to Poland   
4      55354053  Специалист службы поддержки с техническими зна...   
...         ...                                                ...   
49192  55589514           Специалист по информационным технологиям   
49193  55029668                     Руководитель процессного офиса   
49194  66172263                         Начальник сервисной службы   
49195  66180086                             Senior IOS разработчик   
49196  55547842                                   DataOps Engineer   

                                              key_skills          schedule  \
0      Пользователь ПК\tРабота в команде\tРемонт ноут...       Полный день   
1      Средства криптографической защиты информации\t...       Полный день   
2        Spring Framework\tSQL\tHibernate ORM\tJava\tGit  Удаленная работа   
3        Spring Framework\tSQL\tHibernate ORM\tJava\tGit  Удаленная работа   
4                                                   None  Удаленная работа   
...                                                  ...               ...   
49192  Электронный документооборот\t1С: Предприятие 8...       Полный день   
49193  Оптимизация бизнес-процессов\tСтратегическое м...       Полный день   
49194  Организаторские навыки\tУправление персоналом\...       Полный день   
49195  Swift\tRealm\tiOS\tАнглийский язык\tMVP\tMVVM\...       Полный день   
49196                                               None  Удаленная работа   

               experience           employment  salary_from  salary_to  \
0               Нет опыта     Полная занятость      64000.0        NaN   
1           От 3 до 6 лет     Полная занятость          NaN        NaN   
2           От 3 до 6 лет     Полная занятость          NaN        NaN   
3           От 3 до 6 лет     Полная занятость          NaN        NaN   
4               Нет опыта  Частичная занятость      15000.0        NaN   
...                   ...                  ...          ...        ...   
49192       От 3 до 6 лет     Полная занятость          NaN        NaN   
49193       От 3 до 6 лет     Полная занятость          NaN        NaN   
49194  От 1 года до 3 лет     Полная занятость      60000.0        NaN   
49195  От 1 года до 3 лет     Полная занятость     146456.0   244094.0   
49196       От 3 до 6 лет     Полная занятость          NaN        NaN   

       area_id  employer_id  
0         1516      5724811  
1          130      4903713  
2          160        69961  
3          159        69961  
4         1955         1740  
...        ...          ...  
49192       66      5109332  
49193       26        51333  
49194     1930      4094700  
49195      160      4469939  
49196      160      4978627  

[49197 rows x 10 columns]

3.1Сколько вакансий есть в базе? 

In [47]:
query = f'''SELECT COUNT(*)
            FROM public.vacancies    
        '''
df = pd.read_sql_query(query, connection)
df

C:\Users\Polta\AppData\Local\Temp\ipykernel_15108\3903806583.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0  49197

результат запроса:
в базе 49197 вакансий

3.2 Сколько работодателей?

In [48]:
query = f'''SELECT COUNT(*)
            FROM public.employers   
        '''
df = pd.read_sql_query(query, connection)
df

C:\Users\Polta\AppData\Local\Temp\ipykernel_15108\1362427963.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0  23501

результат запроса:в базе 23501 работодателей

3.3 Сколько регионов?

In [49]:
query = f'''SELECT COUNT(distinct id)
            FROM public.areas 
        '''
df = pd.read_sql_query(query, connection)
df

C:\Users\Polta\AppData\Local\Temp\ipykernel_15108\2943110102.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0   1362

результат запроса:
в базе 1362 регионов

3.4 Сколько сфер деятельности?

In [50]:
query = f'''SELECT COUNT(distinct id)
            FROM public.industries        
        '''
df = pd.read_sql_query(query, connection)
df

C:\Users\Polta\AppData\Local\Temp\ipykernel_15108\614354148.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0    294

результат запроса:сфер деятельности в базе 294

4.1 Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).

Выберите пятёрку лидеров по количеству вакансий:

In [51]:
query = f'''SELECT 
                count(v.id),
                a.name              
            FROM public.areas a
                LEFT JOIN public.vacancies v ON v.area_id = a.id                 
            GROUP BY a.id   
            ORDER BY 1 DESC  
        ''' 
df = pd.read_sql_query(query, connection)
df.head(6)

C:\Users\Polta\AppData\Local\Temp\ipykernel_15108\2197013071.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count             name
0   5333           Москва
1   2851  Санкт-Петербург
2   2112            Минск
3   2006      Новосибирск
4   1892           Алматы
5   1698     Екатеринбург


Пятёрка лидеров по количеству вакансий:Москва, Санкт-Петербург, Минск, Новосибирс, Алматы

4.2 Посмотрим на зарплаты. У какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой?

In [52]:
query = f'''SELECT 
                count(id)
            FROM public.vacancies v
            WHERE salary_from IS NOT NULL
                OR salary_to IS NOT NULL
        ''' 
df = pd.read_sql_query(query, connection)
df.iloc[0]

C:\Users\Polta\AppData\Local\Temp\ipykernel_15108\3258602998.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count    24073
Name: 0, dtype: int64

Поле с зарплатой заполнено у 24073 вакансий

4.3 Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого числа.
Нижняя граница:

Верхняя граница:

In [53]:
query = f'''SELECT 
                AVG(salary_from) as from,
                AVG(salary_to) as to
            FROM public.vacancies v
        ''' 
df = pd.read_sql_query(query, connection)
print(f' from {round(df.iloc[0,0])} to {round(df.iloc[0,1])}')

 from 71065 to 110537


C:\Users\Polta\AppData\Local\Temp\ipykernel_15108\3159080043.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


Средние значения для нижней и верхней границы зарплатной вилки: 71065 и 110537

4.4 Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Какая пара находится на втором месте по популярности?

In [54]:
query = f'''SELECT distinct
                schedule,
                employment,
                count(id)
            FROM public.vacancies
            GROUP BY schedule, employment
            ORDER BY 3 DESC
        ''' 
df = pd.read_sql_query(query, connection)
df

C:\Users\Polta\AppData\Local\Temp\ipykernel_15108\2175723460.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

Наиболее популярные сочетания типа рабочего графика и типа трудоустройства: 
Удаленная работа + Полная занятость, Гибкий график + Полная занятость, Удаленная работа + Частичная занятость	

4.5 Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта.

In [55]:
query = f'''SELECT
                experience, 
                count(id)
            FROM public.vacancies
            GROUP BY experience
            ORDER BY 2 ASC
        ''' 
df = pd.read_sql_query(query, connection)
df

C:\Users\Polta\AppData\Local\Temp\ipykernel_15108\1132473073.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

Наибольшее количество вакансий (26152) рассчитано на людей с опытом работы от 1 года до 3 лет, 
на втором месте (14511) - от 3 до 6 лет, 
на третьем (7197) - без опыта, 
на последнем (1337) - с опытом более 6 лет.


5.1 Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [56]:
query = f'''
select 
e.name,
count(v.name)
from vacancies v 
join employers e on v.employer_id = e.id
group by e.name
order by 2 desc
limit 5
        ''' 
df = pd.read_sql_query(query, connection)
print(df)


C:\Users\Polta\AppData\Local\Temp\ipykernel_15108\3961750239.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


            name  count
0         Яндекс   1933
1     Ростелеком    491
2       Тинькофф    444
3           СБЕР    428
4  Газпром нефть    331


5.2
Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.

Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей. Впишите его название в поле ниже в том виде, который вернул запрос.

In [57]:
query = f'''SELECT
                a.name, 
                count(e.id) "Количество работодателей",
                count(v.id) "Количество вакансий"
            FROM public.areas a
                LEFT JOIN public.employers e ON a.id = e.area
                LEFT JOIN public.vacancies v ON v.area_id = a.id
            GROUP BY a.name
            ORDER BY 3 ASC, 2 DESC
        ''' 
df = pd.read_sql_query(query, connection)
print(df)
df.iloc[0,0]

C:\Users\Polta\AppData\Local\Temp\ipykernel_15108\1872406147.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


                    name  Количество работодателей  Количество вакансий
0                 Россия                       410                    0
1              Казахстан                       207                    0
2     Московская область                        75                    0
3     Краснодарский край                        19                    0
4     Ростовская область                        18                    0
...                  ...                       ...                  ...
1357         Новосибирск                   1149438              1149438
1358              Алматы                   1364132              1364132
1359               Минск                   2354880              2354880
1360     Санкт-Петербург                   6320667              6320667
1361              Москва                  31272712             31272712

[1362 rows x 3 columns]


'Россия'

5.3
Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии.
Выберите максимальное значение из получившегося списка.

In [58]:
query = f'''SELECT
                e.name, 
                count(distinct v.area_id) "Количество регионов"
            FROM public.employers e 
                JOIN public.vacancies v ON e.id = v.employer_id
            GROUP BY e.id
            ORDER BY 2 DESC

        ''' 
df = pd.read_sql_query(query, connection)
df

C:\Users\Polta\AppData\Local\Temp\ipykernel_15108\3517608302.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


name  Количество регионов
0                                      Яндекс                  181
1                                  Ростелеком                  152
2                                  Спецремонт                  116
3                      Поляков Денис Иванович                   88
4                                   ООО ЕФИН                    71
...                                       ...                  ...
14901                            НПП Авиатрон                    1
14902              Центр дистанционных торгов                    1
14903  Городские Телекоммуникационные Системы                    1
14904                       Введенский, Отель                    1
14905              Группа Компаний Автоимпорт                    1

[14906 rows x 2 columns]

Задание 5.4
Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности.
Введите количество, которое вернул запрос, в поле ниже.

In [59]:
query = f'''SELECT COUNT(*)
            FROM public.employers e 
                LEFT JOIN public.employers_industries ei ON ei.employer_id = e.id
            WHERE ei.industry_id IS NULL
        ''' 
df = pd.read_sql_query(query, connection)
df.iloc[0]

C:\Users\Polta\AppData\Local\Temp\ipykernel_15108\3656247542.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count    8419
Name: 0, dtype: int64

Задание 5.5
Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности.
Введите в поле ниже название этой компании так же, как оно указано в результате запроса.

In [60]:
query = f'''SELECT e.name "имя работодателя"
            FROM public.employers_industries ei
                JOIN public.employers e ON ei.employer_id = e.id
            GROUP BY e.id --ei.industry_id
            HAVING COUNT(distinct ei.industry_id) = 4
            ORDER BY e.name
            OFFSET 2 LIMIT 1
        ''' 
df = pd.read_sql_query(query, connection)
df.iloc[0]

C:\Users\Polta\AppData\Local\Temp\ipykernel_15108\4259496213.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


имя работодателя    2ГИС
Name: 0, dtype: object

Задание 5.6
С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана «Разработка программного обеспечения».

In [61]:
query = f'''SELECT COUNT(distinct ei.employer_id)
            FROM public.employers_industries ei
                JOIN public.industries i ON ei.industry_id = i.id
            WHERE i.name = 'Разработка программного обеспечения'
            
        ''' 
df = pd.read_sql_query(query, connection)
df.iloc[0]

C:\Users\Polta\AppData\Local\Temp\ipykernel_15108\1655995049.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count    3553
Name: 0, dtype: int64

Задание 5.7
Для компании «Яндекс» выведите список регионов-миллионников , в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании, собранных в этой таблице. Должна получиться выборка такого вида

In [62]:
big_city = (
    "Москва", "Санкт-Петербург", "Новосибирск", 
    "Екатеринбург", "Казань", "Нижний Новгород", 
    "Красноярск", "Челябинск", "Самара", "Уфа", 
    "Ростов-на-Дону", "Омск", "Краснодар", 
    "Воронеж", "Пермь", "Волгоград"
)
query = f'''(SELECT 
                a.name "Город",
                COUNT(v.id) "Количество вакансий в городе"
            FROM public.vacancies v
                JOIN public.employers e ON e.id = v.employer_id
                JOIN public.areas a ON a.id = v.area_id
            WHERE e.name = 'Яндекс' and a.name in {big_city}
            GROUP BY a.id
            ORDER BY 2 ASC)
            
            UNION ALL 

            (SELECT 
                'TOTAL',
                COUNT (*) "Общ кол-во вакансий в городах-миллионниках"
            FROM public.vacancies v
                JOIN public.employers e ON e.id = v.employer_id
                JOIN public.areas a ON a.id = v.area_id
            WHERE e.name = 'Яндекс' and a.name in {big_city})            
           
        ''' 
df = pd.read_sql_query(query, connection)
df

C:\Users\Polta\AppData\Local\Temp\ipykernel_15108\1548986203.py:29: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


Город  Количество вакансий в городе
0             TOTAL                           485
1              Омск                            21
2         Челябинск                            22
3        Красноярск                            23
4         Волгоград                            24
5    Ростов-на-Дону                            25
6            Казань                            25
7             Пермь                            25
8            Самара                            26
9               Уфа                            26
10        Краснодар                            30
11          Воронеж                            32
12      Новосибирск                            35
13  Нижний Новгород                            36
14     Екатеринбург                            39
15  Санкт-Петербург                            42
16           Москва                            54

# выводы по анализу работодателей
Мы посчитали, сколько вакансий открыто у различных работодателей,
для каждого региона вывели количество работодателей и вакансий в нем,
для каждого работодателя посчитали количество регионов, в которых у него есть вакансии,
определили количество работодателей, у которых сфера деятельности не указана, 
а также нашли работодателей с определенным количеством/названиями сфер деятельности,
вывели распределение количества вакансий конкретного работодателя по городам-миллионникам.

Задание 6.1
Сколько вакансий имеет отношение к данным?

In [63]:
query = f'''SELECT COUNT(id)
            FROM public.vacancies
            WHERE    
                lower(name) like '%data%'
                OR lower(name) like '%данн%'
        ''' 
df = pd.read_sql_query(query, connection)
df

C:\Users\Polta\AppData\Local\Temp\ipykernel_15108\2973367864.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0   1771

Задание 6.2
Сколько есть подходящих вакансий для начинающего дата-сайентиста?
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
‘data scientist’;
‘data science’;
‘исследователь данных’;
‘ML’ (здесь не нужно брать вакансии по HTML);
‘machine learning’;
‘машинн%обучен%’.
Считаем вакансиями для специалистов уровня Junior следующие:
в названии есть слово “junior” или
требуемый опыт — «Нет опыта» или
тип трудоустройства — «Стажировка».
В качестве ответа запишите количество вакансий, которое вернул запрос.

In [64]:
query = f'''SELECT COUNT(id)
            FROM public.vacancies
            WHERE (lower(name) like '%data scientist%'
                OR lower(name) like '%data science%'
                OR lower(name) like '%исследователь данных%'
                OR (name like '%ML%' AND lower(name) not like '%html%')
                OR lower(name) like '%machine learning%' 
                OR lower(name) like '%машинн%обучен%')
                AND (lower(name) like '%junior%'
                   OR experience = 'Нет опыта'
                   OR employment = 'Стажировка')
        ''' 
df = pd.read_sql_query(query, connection)
df

C:\Users\Polta\AppData\Local\Temp\ipykernel_15108\148676335.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0     51

Задание 6.3
Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?
В качестве ответа запишите количество, которое вернул запрос.

In [65]:
query = f'''SELECT COUNT(id)
            FROM public.vacancies
            WHERE ((lower(name) like '%data scientist%'
                OR lower(name) like '%data science%'
                OR lower(name) like '%исследователь данных%'
                OR (name like '%ML%' AND lower(name) not like '%html%')
                OR lower(name) like '%machine learning%' 
                OR lower(name) like '%машинн%обучен%'))
                AND (lower(key_skills) like '%sql%' 
                    OR lower(key_skills) like '%postgres%')
        ''' 
df = pd.read_sql_query(query, connection)
df

C:\Users\Polta\AppData\Local\Temp\ipykernel_15108\2226981322.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0    201

Задание 6.4
С помощью запроса, аналогичного предыдущему, проверьте, насколько популярен Python в требованиях работодателей к DS. Вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.
В качестве ответа запишите количество, которое вернул запрос.

In [66]:
query = f'''select count(*)
    from
        public.vacancies v
    where
        (
        name ilike '%data scientist%'  or
        name ilike  '%data science%' or
        name ilike  '%исследователь данных%'  or
        (name ilike  '%ML%' and name not ilike '%HTML%')  or
        name ilike '%machine learning%' or  
        name ilike '%машинн%обучен%'
        )
        and
        key_skills ilike '%Python%'
        ''' 
df = pd.read_sql_query(query, connection)
df

C:\Users\Polta\AppData\Local\Temp\ipykernel_15108\854745057.py:16: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0    357

Задание 6.5
Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [67]:
query = f'''SELECT 
                ROUND(AVG(
                    length(key_skills) + 1 - length(REPLACE(key_skills, CHR(9), ''))
                ), 2) 
            FROM public.vacancies 
            WHERE ((lower(name) like '%data scientist%'
                OR lower(name) like '%data science%'
                OR lower(name) like '%исследователь данных%'
                OR (name like '%ML%' AND lower(name) not like '%html%')
                OR lower(name) like '%machine learning%' 
                OR lower(name) like '%машинн%обучен%'))
                AND (key_skills IS NOT NULL)
        ''' 
df = pd.read_sql_query(query, connection)
df.iloc[0]

C:\Users\Polta\AppData\Local\Temp\ipykernel_15108\149360332.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


round    6.41
Name: 0, dtype: float64

Задание 6.6
Напишите запрос, позволяющий вычислить, какую зарплату для DS в среднем указывают для каждого типа требуемого опыта (уникальное значение из поля experience).
Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа.

In [68]:
query = f'''
select

  v.experience, ROUND(AVG(coalesce((v.salary_from + v.salary_to) / 2, v.salary_from, v.salary_to)), 0)

from vacancies v

WHERE
(
       (v.name ILIKE '%data scientist%')
    OR (lower(v.name) ILIKE '%data science%')
    OR (v.name ILIKE '%исследователь данных%')
    OR (v.name  LIKE '%ML%' AND  (v.name NOT iLIKE '%html%'))
    OR (v.name ILIKE '%machine learning%')
    OR (v.name ILIKE '%машинн%обучен%')
    AND (key_skills IS NOT NULL)
)

GROUP BY v.experience
ORDER BY 1      
        ''' 
df = pd.read_sql_query(query, connection)
df

C:\Users\Polta\AppData\Local\Temp\ipykernel_15108\1185194429.py:22: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


experience     round
0         Более 6 лет       NaN
1           Нет опыта   74643.0
2  От 1 года до 3 лет  145236.0
3       От 3 до 6 лет  243115.0

# выводы по предметному анализу
В этом разделе мы нашли вакансии, связанные с данными, а также вакансии конкретно для дата-сайентистов.
Среди последних посчитали количество тех, которые подходят начинающим специалистам, 
а также количество вакансий, в которых в качестве ключевых навыков указаны SQL/Postrges или Python, 
посчитали среднее количество требуемых ключевых навыков,
определили среднюю зарплату в зависимости от опыта.

В нашей базе данных содержатся вакансии из различных сфер деятельностей, стран, регионов, с различными требуемыми навыками, опытом работы, условиями работы и зарплатой. Наибольшее количество вакансий нашей базы данных находится в столицах (как России, так и других стран) и больших городах, в небольших населённых пунктах разнообразие вакансий гораздо меньше. Максимальное количество вакансий соответствует сфере деятельности "Разработка программного обеспечения". Это можно увидеть из следующего запроса:

In [69]:
query = f'''SELECT 
                i.name,
                count(v.id)
            FROM 
                public.industries i
                JOIN public.employers_industries ei ON ei.industry_id = i.id
                JOIN public.employers e ON ei.employer_id = e.id
                JOIN public.vacancies v ON v.employer_id = e.id
            GROUP BY i.name 
            ORDER BY 2 DESC        
        ''' 
df = pd.read_sql_query(query, connection)
df

C:\Users\Polta\AppData\Local\Temp\ipykernel_15108\3781892273.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


name  count
0                  Разработка программного обеспечения  12499
1    Системная интеграция,  автоматизации технологи...  11034
2    Интернет-компания (поисковики, платежные систе...   6413
3                                                 Банк   2742
4                                      Мобильная связь   1585
..                                                 ...    ...
288                               Спортивная федерация      5
289                                          Химчистка      4
290             Ассоциация в сфере культуры, искусства      2
291                                          Автошкола      1
292                                  Ритуальные услуги      1

[293 rows x 2 columns]

Также можно отметить, что специалисты, чья работа связана с данными, больше всего нужны в таких сферах, как "Разработка программного обеспечения", "Системная интеграция, автоматизации технологических и бизнес-процессов предприятия, ИТ-консалтинг" и "Банк"

In [70]:
query = f'''SELECT 
                i.name,
                count(v.id)
            FROM 
                public.industries i
                JOIN public.employers_industries ei ON ei.industry_id = i.id
                JOIN public.employers e ON ei.employer_id = e.id
                JOIN public.vacancies v ON v.employer_id = e.id
            WHERE     
                (lower(v.name) like '%data%'
                OR lower(v.name) like '%данн%')
            GROUP BY i.name 
            ORDER BY 2 DESC       
        ''' 
df = pd.read_sql_query(query, connection)
df

C:\Users\Polta\AppData\Local\Temp\ipykernel_15108\421976542.py:15: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


name  count
0                  Разработка программного обеспечения    514
1    Системная интеграция,  автоматизации технологи...    421
2                                                 Банк    293
3    Интернет-компания (поисковики, платежные систе...    215
4                                      Мобильная связь    137
..                                                 ...    ...
187         Сантехника (продвижение, оптовая торговля)      1
188  Неорганическая химия (продвижение, оптовая тор...      1
189  Оборудование для ресторанов, гостиниц, кейтери...      1
190         Резина, пластмасса и прочее (производство)      1
191  Консервированная продукция (продвижение, оптов...      1

[192 rows x 2 columns]

Что же касается вакансий конкретно для DS, то наибольшее их количество требуется всё в тех же трёх сферах, что и в предыдущем запросе. Похожие результаты наблюдаются и для новичков-DS, причем в вакансиях для них среди требуемых ключевых навыков чаще других встречается Python, SQL, английский язык, Machine Learning, математический анализ и математическая статистика, зарплата находится в диапазоне от 45 до 150 тысяч рублей, а среди типов трудоустройства самым широко встречающимся является полная занятость, среди типов рабочего графика - полный день

In [71]:
query = f'''SELECT i.name,
                   COUNT(v.id)
            FROM public.industries i
                JOIN public.employers_industries ei ON ei.industry_id = i.id
                JOIN public.employers e ON ei.employer_id = e.id
                JOIN public.vacancies v ON v.employer_id = e.id
            WHERE (lower(v.name) like '%data scientist%'
                OR lower(v.name) like '%data science%'
                OR lower(v.name) like '%исследователь данных%'
                OR (v.name like '%ML%' AND lower(v.name) not like '%html%')
                OR lower(v.name) like '%machine learning%' 
                OR lower(v.name) like '%машинн%обучен%')
                AND (lower(v.name) like '%junior%'
                OR v.experience = 'Нет опыта'
                OR v.employment = 'Стажировка')
            GROUP BY i.name  
            ORDER BY 2 DESC
        ''' 
df = pd.read_sql_query(query, connection)
df

C:\Users\Polta\AppData\Local\Temp\ipykernel_15108\3540538653.py:19: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


name  count
0                 Разработка программного обеспечения     16
1                                                Банк     13
2   Управляющая, инвестиционная компания (управлен...     10
3   Системная интеграция,  автоматизации технологи...      8
4   Интернет-компания (поисковики, платежные систе...      7
5   Научно-исследовательская, научная, академическ...      5
6                               Консалтинговые услуги      4
7   Электронно-вычислительная, оптическая, контрол...      3
8   Бытовая техника, электроника, климатическое об...      2
9                         Государственные организации      2
10  Розничная сеть (drogerie, товары повседневного...      2
11            Автокомпоненты, запчасти (производство)      2
12  Неорганическая химия (продвижение, оптовая тор...      1
13                Неорганическая химия (производство)      1
14  Авиационная, вертолетная, аэрокосмическая пром...      1
15         Нефтехимия (продвижение, оптовая торговля)      1
16  Промышленное, бытовое электрооборудование и эл...      1
17  Промышленное, бытовое электрооборудование и эл...      1
18  Промышленное, бытовое электрооборудование и эл...      1
19             Розничная сеть (DIY и товары для дома)      1
20         Розничная сеть (одежда, обувь, аксессуары)      1
21                       Розничная сеть (продуктовая)      1
22                                   Складские услуги      1
23                       Страхование, перестрахование      1
24  Стройматериалы, сборные конструкции для ремонт...      1
25                             Туристические компании      1
26                                Фиксированная связь      1
27  Финансово-кредитное посредничество (биржа, бро...      1
28  Электронно-вычислительная, оптическая, контрол...      1
29                           Юридическое обслуживание      1
30        Нефтепереработка, нефтехимия (производство)      1
31  Бытовая химия, парфюмерия, косметика  (продвиж...      1
32  Бытовая химия, парфюмерия, косметика  (произво...      1
33                             Вуз, ссуз колледж, ПТУ      1
34                  ГСМ, топливо (розничная торговля)      1
35                                        Добыча газа      1
36                                       Добыча нефти      1
37                                   Интернет-магазин      1
38  Контроль качества, экспертиза, испытания и сер...      1
39                      Курьерская, почтовая доставка      1
40  Маркетинговые, рекламные, BTL, дизайнерские, E...      1
41                                    Мобильная связь      1

In [72]:

#ключевые навыки для новичков DS
query = f'''SELECT key_skills
            FROM public.vacancies v
            WHERE (lower(v.name) like '%data scientist%'
                OR lower(v.name) like '%data science%'
                OR lower(v.name) like '%исследователь данных%'
                OR (v.name like '%ML%' AND lower(v.name) not like '%html%')
                OR lower(v.name) like '%machine learning%' 
                OR lower(v.name) like '%машинн%обучен%')
                
                AND (lower(v.name) like '%junior%'
                OR v.experience = 'Нет опыта'
                OR v.employment = 'Стажировка')
            ORDER BY 1
        ''' 
df = pd.read_sql_query(query, connection)
df

C:\Users\Polta\AppData\Local\Temp\ipykernel_15108\3968133373.py:16: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


key_skills
0   Computer Vision\tPyTorch\tTensorflow\tDeep Lea...
1                                       Linux\tPython
2                                       Linux\tPython
3   Machine Learning\tDeep Learning\tData Science\...
4   Machine Learning\tDeep Learning\tData Science\...
5   Machine Learning\tDeep Learning\tData Science\...
6   ML\tлинейные модели\tRandom Forest\tGradient B...
7   Python\tC++\tОптимизация кода\tАнализ кода\tМа...
8          Python\tDjango Framework\tSQL\tXML\tMS SQL
9   Python\tGit\tTensorFlow\tKeras\tJupiter\tNLP\t...
10  Python\tLinux\tGit\tSQL\tDocker\tBash\tOpenCV\...
11  Python\tMachine Learning\tPandas\tSQL\tNumpy\t...
12  Python\tMachine Learning\tМатематическое модел...
13  Python\tMATLAB\tМатематическое моделирование\t...
14                                         Python\tML
15               Python\tML\tPytorch\tComputer Vision
16  Python\tOpenCV\tTensorflow\tComputer Vision\tН...
17  Python\tPyTorch\tGit\tTensorflow\tMachine Lear...
18  Python\tSQL\tData Mining\tC++\tМатематический ...
19  Python\tSQL\tJava\tМатематический анализ\tМате...
20  Python\tSQL\tJava\tМатематический анализ\tМате...
21  Python\tSQL\tMachine Learning\tData Science\tА...
22       Python\tSQL\tMachine Learning\tDocker\tNumpy
23  Python\tSQL\tORACLE\tСтатистический анализ\tSp...
24  Python\tSQL\tPostgreSQL\tML\tMachine Learning\...
25           Python\tSQL\tАнглийский язык\tLinux\tООП
26  Python\tSQL\tАнглийский язык\tОбучение и разви...
27          Python\tSQL\tРабота в команде\tOpenCV\tML
28  Python\tTensorflow\tPyTorch\tMachine Learning\...
29  Python\tUnix\tАналитическое мышление\tНейронны...
30  Python\tUnix\tАналитическое мышление\tНейронны...
31  Python\tМатематическая статистика\tOracle Pl/S...
32   Python\tМатематическая статистика\tАнализ данных
33  Python\tМатематическая статистика\tПрогнозиров...
34  PyTorch\tMachine Learning\tКонференции\tPython...
35  SQL\tMS SQL\tУдаленная работа\tPython\tМатемат...
36                                 SQL\tOracle Pl/SQL
37  Английский язык\tCJM\tПодготовка презентаций\t...
38  Математический анализ\tБазы данных\tСтатистика...
39  Математическое моделирование\tАнализ данных\tС...
40  Оценка рисков\tMS PowerPoint\tСтатистический а...
41  Оценка рисков\tMS PowerPoint\tСтатистический а...
42  Оценка рисков\tMS PowerPoint\tСтатистический а...
43  Оценка рисков\tMS PowerPoint\tСтатистический а...
44  Оценка рисков\tMS PowerPoint\tСтатистический а...
45  Оценка рисков\tMS PowerPoint\tСтатистический а...
46  Оценка рисков\tMS PowerPoint\tСтатистический а...
47  Оценка рисков\tMS PowerPoint\tСтатистический а...
48  Оценка рисков\tMS PowerPoint\tСтатистический а...
49                                               None
50                                               None

In [73]:
#предельные значения зарплаты новичков DS
query = f'''SELECT MIN(salary_from), MAX(salary_to)
            FROM public.vacancies v
            WHERE (lower(v.name) like '%data scientist%'
                OR lower(v.name) like '%data science%'
                OR lower(v.name) like '%исследователь данных%'
                OR (v.name like '%ML%' AND lower(v.name) not like '%html%')
                OR lower(v.name) like '%machine learning%' 
                OR lower(v.name) like '%машинн%обучен%')
                
                AND (lower(v.name) like '%junior%'
                OR v.experience = 'Нет опыта'
                OR v.employment = 'Стажировка')
        ''' 
df = pd.read_sql_query(query, connection)
print(f'Salary from {round(df.iloc[0,0])} to {round(df.iloc[0,1])}')

Salary from 45000 to 150000


C:\Users\Polta\AppData\Local\Temp\ipykernel_15108\321009366.py:15: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


In [74]:
#определяем самый популярный тип трудоустройства и тип рабочего графика для новичков DS
query = f'''(SELECT 
                'тип трудоустройства',
                employment,
                COUNT(employment)
            FROM public.vacancies v
            WHERE (lower(v.name) like '%data scientist%'
                OR lower(v.name) like '%data science%'
                OR lower(v.name) like '%исследователь данных%'
                OR (v.name like '%ML%' AND lower(v.name) not like '%html%')
                OR lower(v.name) like '%machine learning%' 
                OR lower(v.name) like '%машинн%обучен%')
                
                AND (lower(v.name) like '%junior%'
                OR v.experience = 'Нет опыта'
                OR v.employment = 'Стажировка')
            GROUP BY employment
            ORDER BY 3 DESC
            LIMIT 1)

            UNION ALL
                
            (SELECT 
                'тип рабочего графика',
                schedule,
                COUNT(schedule)
            FROM public.vacancies v
            WHERE (lower(v.name) like '%data scientist%'
                OR lower(v.name) like '%data science%'
                OR lower(v.name) like '%исследователь данных%'
                OR (v.name like '%ML%' AND lower(v.name) not like '%html%')
                OR lower(v.name) like '%machine learning%' 
                OR lower(v.name) like '%машинн%обучен%')
                
                AND (lower(v.name) like '%junior%'
                OR v.experience = 'Нет опыта'
                OR v.employment = 'Стажировка')  
            GROUP BY schedule
            ORDER BY 3 DESC
            LIMIT 1)
        ''' 
df = pd.read_sql_query(query, connection)
df

C:\Users\Polta\AppData\Local\Temp\ipykernel_15108\1831172557.py:42: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


?column?        employment  count
0   тип трудоустройства  Полная занятость     42
1  тип рабочего графика       Полный день     36

In [75]:
connection.close()